src/utils/work/seekWorkResponse.js

```diff
@@ -146,30 +134,43 @@ const dispatchWorkRequest = async (
   });
 
   const responsePromise = new Promise((resolve, reject) => {
-    io.on(`WorkResponse-${ETag}`, (res) => {
-      const { response } = res;
-
-      if (response.error) {
-        const { errorCode, userMessage } = response;
-        console.error(errorCode, userMessage);
-
-        return reject(
-          new WorkResponseError(errorCode, userMessage, request),
-        );
+    io.on(`WorkResponse-${ETag}`, async (res) => {
+      // If type is object, then we received a notification with a signedUrl
+      // now we need to fetch the actual result from s3
+      if (typeof res === 'object') {
+        const { response } = res;
+
+        if (response.error) {
+          const { errorCode, userMessage } = response;
+          console.error(errorCode, userMessage);
+
+          return reject(
+            new WorkResponseError(errorCode, userMessage, request),
+          );
+        }
+
+        return resolve({ signedUrl: response.signedUrl });
       }
 
-      // If no error, the response should be ready on S3.
-      // In this case, return true
-      return resolve();
+      // If type isn't object, then we have the actual work result,
+      // no further downloads are necessary, we just need to decompress and return it
+      const decompressedData = await decompressUint8Array(Uint8Array.from(Buffer.from(res, 'base64')));
+      return resolve({ data: parseResult(decompressedData) });
     });
   });
 
-  const result = Promise.race([timeoutPromise, responsePromise]);
+  // TODO test what happens when api throws an error here
+  await fetchAPI(
+    `/v2/workRequest/${experimentId}/${ETag}`,
+    {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(request),
+    },
+  );
```

Comment:
I joined the roles of the socket WorkRequest-v2 and `workResults/` http endpoint into one single endpoint `workRequest`.

This endpoint checks if the result is cached in S3
  - If it is, it returns over the `WorkResults-{ETag}` the signedUrl.
  - If it isn't, it submits the work request

Reply:
I believe this is better than the previous situation.

---

The `WorkRequest-v2` didn't need to be a socket and made the UI a bit more complicated handling different ways of contacting the api. I would consider using a socket (redis adapter) though if we somehow used it to connect directly to the worker to avoid the api as a middleman, but the danger there is that users can send whatever trash they want directly to the R worker, a potential vulnerability. Also I tried this out in staging and didn't notice a significant improvement (the api and worker are in the same region, so added delay is relatively insignificant)

---

Using `workResults/` meant that: 

**If the result is cached in S3** we perform 1 call to the api and wait for the result. This hasn't changed with the new method, only that now we wait for the result in a socket instead of the http request itself.



**If the result isn't cached in S3,** we have 4 back and forths with the server in our aws region:
- one call to the api, it fails (results aren't in S3)
- one call to the api (submit work request)
- one call to the api to get the signedUrl (to fetch the results we were notified are now in s3)- one call to S3 to download with the signedUrl

Now instead we just perform either 2 calls:
- one call to the api to trigger workRequest
- one call to S3 to download with the signedUrl we received in the socket (the work result is too big to send over a socket fast)

or 1 call:
- one call to the api to trigger workRequest
- we received the result in the socket, nothing else to do

Reply:
have you actually tested whether workRequest + S3 is slower than workRequest + socket? I've seen comments on fine-tuning in other PRs so I assume you've been running some tests to compare both. It'd be good to see those numbers because upfront I'm not very convinced on the extra complexity of sending through sockets. 

Modifying the current code to have a single call is a great idea. That single call could also return the `ETag` right? so it wouldn't conflict with the refactor to move the `ETag` compute to the API.

Reply:
Finished testing, posted the results [here](https://this-is-biomage.slack.com/archives/C015CHSUDRU/p1694788525718699).

There is a google doc in that message with more details about the tests